src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/linalg/IterativeLinearSolverNDArray.java

```diff
@@ -0,0 +1,183 @@
+    private final Function<INDArray, Double> normFunc;
+    private final BiFunction<INDArray, INDArray, Double> innerProductFunc;
+    private final boolean performDetailedChecks;
+    private final double absTol, relTol;
+    private final int maxIters;
+
+    public enum ExitStatus {
+        SUCCESS_ABS_TOL,
+        SUCCESS_REL_TOL,
+        FAIL_MAX_ITERS
+    }
+
+    /**
+     * Constructor
+     *
+     * @param linop a linear operator (a rank-2 matrix or an abstract operator)
+     * @param b rhs vector (a rank-1 object)
+     * @param precond the preconditioner
+     * @param absTol absolute tolerance
+     * @param relTol relative tolerance
+     * @param maxIters maximum iterations
+     * @param normFunc a norm function
+     * @param innerProductFunc an inner product function
+     * @param performDetailedChecks perform detailed checks
+     */
+    public IterativeLinearSolverNDArray(@Nonnull final GeneralLinearOperator<INDArray> linop,
+                                        @Nonnull final INDArray b,
+                                        @Nullable final GeneralLinearOperator<INDArray> precond,
+                                        final double absTol, final double relTol,
+                                        final int maxIters,
+                                        @Nonnull final Function<INDArray, Double> normFunc,
+                                        @Nonnull final BiFunction<INDArray, INDArray, Double> innerProductFunc,
+                                        final boolean performDetailedChecks) {
+        this.performDetailedChecks = performDetailedChecks;
+        this.linop = linop;
+        if (b.length() != linop.getColumnDimension()) {
+            throw new IllegalArgumentException("The provided rhs vector has wrong dimensions");
+        } else {
+            this.b = b;
+        }
+
+        this.precond = precond;
+        if (precond != null) {
+            if (linop.getColumnDimension() != precond.getColumnDimension() ||
+                    linop.getRowDimension() != precond.getRowDimension()) {
+                throw new IllegalArgumentException("The preconditioner and linear operators must have the same " +
+                        "row and column dimensions");
+            }
+        }
+
+        this.normFunc = normFunc;
+        this.innerProductFunc = innerProductFunc;
+
+        this.absTol = ParamUtils.isPositive(absTol, "Absolute error tolerance must be positive");
+        this.relTol = ParamUtils.isPositive(relTol, "Relative error tolerance must be pisitive");
+        this.maxIters = ParamUtils.isPositive(maxIters, "Maximum iterations must be positive");
+    }
+
+    /**
+     * Preconditioned conjugate gradients method
+     *
+     * @param x0 initial guess
+     * @return an instance of {@link SubroutineSignal}
+     * @throws IllegalArgumentException for bad intial guess or illegla linear operators
+     */
+    public SubroutineSignal cg(@Nonnull final INDArray x0)
```

Comment:
This is quite general. `SubroutineSignal` and `StandardSubroutineSignals` are both public (the latter is introduced in recent commits). Perhaps you mean we should move these out of coveragemodel package, in which case, I totally agree; but then again, let's do repackaging later.